API/c/Src/analog_in.c

```diff
@@ -0,0 +1,65 @@
+#include "analog_in.h"
+
+ADC_HandleTypeDef ADC_InitStruct;
+
+const int REFRESH_PERIOD = 10;
+const int ADC_MAX = 4096;
+const float ADC_RATIO = 3.3 / 4096;
+
+void adc_init(adc_pin_name pin) {
+  __HAL_RCC_ADC3_CLK_ENABLE();
+  GPIO_InitTypeDef GPIO_InitStruct = {};
+  GPIO_InitStruct.Pin = pin.pin;
+  GPIO_InitStruct.Mode = GPIO_MODE_ANALOG;
+  GPIO_InitStruct.Pull  = GPIO_NOPULL;
+  HAL_GPIO_Init(pin.port, &GPIO_InitStruct);
+
+  ADC_InitStruct.Instance                   = pin.adc;
+  ADC_InitStruct.Init.ClockPrescaler        = ADC_CLOCK_SYNC_PCLK_DIV2;
+  ADC_InitStruct.Init.Resolution            = ADC_RESOLUTION_12B;
+  ADC_InitStruct.Init.ScanConvMode          = DISABLE; /* Sequencer disabled (ADC conversion on only 1 channel: channel set on rank 1) */
+  ADC_InitStruct.Init.ContinuousConvMode    = ENABLE; /* Continuous mode enabled to have continuous conversion */
+  ADC_InitStruct.Init.DiscontinuousConvMode = DISABLE; /* Parameter discarded because sequencer is disabled */
+  ADC_InitStruct.Init.NbrOfDiscConversion   = 0;
+  ADC_InitStruct.Init.ExternalTrigConvEdge  = ADC_EXTERNALTRIGCONVEDGE_NONE; /* Conversion start not triggered by external event */
+  ADC_InitStruct.Init.ExternalTrigConv      = ADC_EXTERNALTRIGCONV_T1_CC1;
+  ADC_InitStruct.Init.DataAlign             = ADC_DATAALIGN_RIGHT;
+  ADC_InitStruct.Init.NbrOfConversion       = 1;
+  ADC_InitStruct.Init.DMAContinuousRequests = ENABLE;
+  ADC_InitStruct.Init.EOCSelection          = DISABLE;
+  //initializes the ADC given the structure that defines it
+  if (HAL_ADC_Init(&ADC_InitStruct) != HAL_OK) {
+      //ERROR HANDLING
+  }
+
+  ADC_ChannelConfTypeDef sConfig = {};
+  sConfig.Channel       = pin.channel;
+  sConfig.Rank          = 1;
```

Comment:
Rank has some defines to use instead of constants. See https://github.com/RoboJackets/mtrain-firmware/blob/47e430a8b6b91630540f1171bd006a70f1f1c48e/external/STM32F7xx_HAL_Drivers/Inc/stm32f7xx_hal_adc.h#L354